File: backend/app/services/expense_service.py
from datetime import date

from sqlalchemy.orm import Session

from app.models.expense import Expense
from app.schemas.expense import ExpenseCreate


def create_expense(db: Session, data: ExpenseCreate, created_by: int) -> Expense:
    expense = Expense(**data.model_dump(), created_by=created_by)
    db.add(expense)
    db.commit()
    db.refresh(expense)
    return expense


def list_expenses(
    db: Session,
    date_from: date | None = None,
    date_to: date | None = None,
    offset: int = 0,
    limit: int | None = None,
) -> list[Expense]:
    query = db.query(Expense)
    if date_from is not None:
        query = query.filter(Expense.expense_date >= date_from)
    if date_to is not None:
        query = query.filter(Expense.expense_date <= date_to)
    query = query.order_by(Expense.expense_date.desc(), Expense.id.desc())
    if offset:
        query = query.offset(offset)
    if limit is not None:
        query = query.limit(limit)
    return query.all()

File: frontend/lib/services/expense_service.dart
import '../core/constants/api_constants.dart';
import '../core/network/api_client.dart';
import '../core/utils/date_utils.dart';
import '../models/retail_models.dart';

class ExpenseService {
  const ExpenseService(this._api);

  final ApiClient _api;

  Future<List<Expense>> expenses({
    DateTime? dateFrom,
    DateTime? dateTo,
    int? offset,
    int? limit,
  }) async {
    final data = await _api.get(
      ApiConstants.expenses,
      query: {
        'date_from': dateFrom == null ? null : AppDateUtils.query(dateFrom),
        'date_to': dateTo == null ? null : AppDateUtils.query(dateTo),
        'offset': offset,
        'limit': limit,
      },
    );
    return (data as List)
        .map((item) => Expense.fromJson(Map<String, dynamic>.from(item as Map)))
        .toList();
  }

  Future<Expense> create({
    required String type,
    required double amount,
    required DateTime date,
    String? note,
  }) async {
    final data = Map<String, dynamic>.from(
      await _api.post(
            ApiConstants.expenses,
            data: {
              'expense_type': type,
              'amount': amount,
              'expense_date': AppDateUtils.query(date),
              'note': note,
            },
          )
          as Map,
    );
    return Expense.fromJson(data);
  }
}
